File: src/github_release.rs
use crate::utils::IsBlank;
use crate::CreatedRelease;
use anyhow::{bail, Context, Result};
use serde::{Deserialize, Serialize};

pub(crate) struct GitHubApiClient<'a> {
    api_url: &'a str,
    http_client: reqwest::blocking::Client,
}

impl GitHubApiClient<'_> {
    pub(crate) fn new<'a>(api_url: &'a str) -> GitHubApiClient<'a> {
        GitHubApiClient {
            api_url,
            http_client: reqwest::blocking::Client::new(),
        }
    }

    pub(crate) fn create_github_release(
        &self,
        tag_name: &str,
        release_title: &str,
        changelog: &str,
        repo_name: &str,
        repo_owner: &str,
        auth_token: &str,
    ) -> Result<CreatedRelease> {
        let url = format!("{}/repos/{repo_owner}/{repo_name}/releases", self.api_url);

        let request_body = CreateReleaseRequest {
            tag_name,
            name: release_title,
            body: changelog,
        };

        let response = self
            .http_client
            .post(url)
            .json(&request_body)
            .header("Accept", "application/vnd.github+json")
            .header("Authorization", format!("Bearer {auth_token}"))
            .header("X-GitHub-Api-Version", "2022-11-28")
            // GitHub API requires sending a User-Agent header, and they recommend setting it to
            // your GitHub username:
            // https://docs.github.com/en/rest/using-the-rest-api/getting-started-with-the-rest-api#user-agent
            // In this case, that will be the `repo_owner`.
            .header("User-Agent", repo_owner)
            .send()
            .context("Failed to send create release request to GitHub")?;

        if !response.status().is_success() {
            let response_status = response.status();

            let (body_prefix, response_body): (&str, String) = match response.text() {
                Err(_) => ("and failed to read response body", "".to_owned()),
                Ok(body) if body.is_blank() => ("with blank response body", "".to_owned()),
                Ok(body) => ("response_body:\n", body),
            };
            bail!(
                "Got unsuccessful response ({response_status}) from GitHub when trying to create release, {body_prefix}{response_body}",
            )
        }

        let response_body = response.json::<CreateReleaseResponse>()
            .context("GitHub create release request succeeded, but failed to get release URL from response body")?;

        Ok(CreatedRelease {
            name: release_title.to_owned(),
            url: response_body.html_url,
        })
    }
}

#[derive(Serialize)]
struct CreateReleaseRequest<'a> {
    tag_name: &'a str,
    name: &'a str,
    body: &'a str,
}

#[derive(Deserialize)]
struct CreateReleaseResponse {
    html_url: String,
}

File: internal/changelogrelease/github_release.go
package changelogrelease

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"hermannm.dev/wrap"
	"io"
	"net/http"
)

type GitHubApiClient struct {
	httpClient *http.Client
	apiURL     string
}

func (client GitHubApiClient) createRelease(
	ctx context.Context,
	tagName string,
	releaseTitle string,
	changelog string,
	repoName string,
	repoOwner string,
	authToken string,
) (CreatedRelease, error) {
	url := fmt.Sprintf("%s/repos/%s/%s/releases", client.apiURL, repoOwner, repoName)

	requestBody := CreateReleaseRequest{
		TagName: tagName,
		Name:    releaseTitle,
		Body:    changelog,
	}
	requestBodyJson, err := json.Marshal(requestBody)
	if err != nil {
		return CreatedRelease{}, wrap.Error(
			err,
			"Failed to encode GitHub release request body JSON",
		)
	}

	request, err := http.NewRequestWithContext(
		ctx,
		http.MethodPost,
		url,
		bytes.NewBuffer(requestBodyJson),
	)
	if err != nil {
		return CreatedRelease{}, wrap.Error(err, "Failed to create GitHub release HTTP request")
	}
	request.Header.Add("Authorization", "Bearer "+authToken)
	// GitHub API requires sending a User-Agent header, and they recommend setting it to your GitHub
	// username: https://docs.github.com/en/rest/using-the-rest-api/getting-started-with-the-rest-api#user-agent
	// In this case, that will be the repo owner.
	request.Header.Add("User-Agent", repoOwner)
	request.Header.Add("Accept", "application/vnd.github+json")
	request.Header.Add("X-GitHub-Api-Version", "2022-11-28")

	response, err := http.DefaultClient.Do(request)
	if err != nil {
		return CreatedRelease{}, wrap.Error(err, "Failed to send create release request to GitHub")
	}
	defer response.Body.Close()

	if !isSuccessResponse(response) {
		responseBody := readErrorResponseBody(response)
		// TODO: Replace with ctxwrap.NewErrorWithAttrs with responseStatus, responseBody attrs
		return CreatedRelease{}, fmt.Errorf(
			"Got unsuccessful response from GitHub when trying to create release (responseStatus: %d, responseBody: %s)",
			response.StatusCode, responseBody,
		)
	}

	var responseBody CreateReleaseResponse
	if err := json.NewDecoder(response.Body).Decode(&responseBody); err != nil {
		return CreatedRelease{}, wrap.Error(
			err,
			"GitHub create release request succeeded, but failed to get release URL from response body",
		)
	}

	return CreatedRelease{
		Name: releaseTitle,
		URL:  responseBody.HtmlURL,
	}, nil
}

type CreateReleaseRequest struct {
	TagName string `json:"tag_name"`
	Name    string `json:"name"`
	Body    string `json:"body"`
}

type CreateReleaseResponse struct {
	HtmlURL string `json:"html_url"`
}

func isSuccessResponse(response *http.Response) bool {
	return response.StatusCode >= 200 && response.StatusCode <= 299
}

func readErrorResponseBody(response *http.Response) string {
	responseBody, err := io.ReadAll(response.Body)
	if err != nil {
		return "<failed to read>"
	}
	return string(responseBody)
}
